### PR TITLE
Use the default shutdown flow within extensions

### PR DESCRIPTION
**Pull request by theopolis**
Extension processes should have a way to communicate an exit flow to their service threads. In most cases they will use a similar `Dispatcher`-style service start/stop. This can implement a "core" watcher to shutdown the extension if the osquery core is exited. If the extension opts into this flow it should tear down resources property.

This helps greatly in checking for memory leaks within MSAN.

### CODE REVIEW

**Comment by osqueryer** [comment]
Build origin/pr/1949/merge (Job: 2498) failed :-1:

File: osquery/core/init.cpp [hunk 1]
```diff
@@ -374,7 +388,7 @@ void Initializer::initWatcher() const {
     } else {
       retcode = EXIT_FAILURE;
     }
-    osquery::shutdown(retcode, true);
+    shutdown(retcode);
```
Comment (theopolis) [inline osquery/core/init.cpp]: This should be `::requestShutdown`

File: osquery/core/init.cpp [hunk 2]
```diff
@@ -469,7 +483,7 @@ void Initializer::start() const {
       LOG(ERROR) << RLOG(1629) << binary_
                  << " initialize failed: Could not initialize database";
       auto retcode = (isWorker()) ? EXIT_CATASTROPHIC : EXIT_FAILURE;
-      ::exit(retcode);
+      shutdown(retcode);
```
Comment (theopolis) [inline osquery/core/init.cpp]: This should be `::requestShutdown`

**Comment by osqueryer** [comment]
Build origin/pr/1949/merge (Job: 2504) failed :-1:

File: osquery/events/darwin/scnetwork.cpp [hunk 3]
```diff
@@ -151,6 +151,7 @@ void SCNetworkEventPublisher::stop() {
 }
 
 Status SCNetworkEventPublisher::run() {
+  /*
```
Comment (theopolis) [inline osquery/events/darwin/scnetwork.cpp]: This needs work.

File: tools/tests/test_base.py [hunk 4]
```diff
@@ -420,8 +420,8 @@ class TimeoutRunner(object):
     def __init__(self, cmd=[], timeout_sec=1):
         self.stdout = None
         self.stderr = None
-        self.proc = subprocess.Popen(" ".join(cmd),
-                                     shell=True,
+        self.proc = subprocess.Popen(cmd,
+                                     #shell=True,
```
Comment (theopolis) [inline tools/tests/test_base.py]: Remove this comment.

File: tools/tests/test_osqueryd.py [hunk 5]
```diff
@@ -54,7 +54,7 @@ def info_exists():
         self.assertTrue(os.path.exists(info_path))
         daemon.kill()
 
-    @test_base.flaky
+    #@test_base.flaky
```
Comment (theopolis) [inline tools/tests/test_osqueryd.py]: Restore this.

File: tools/tests/test_osqueryd.py [hunk 6]
```diff
@@ -82,7 +82,7 @@ def test_4_daemon_sighup(self):
         os.kill(daemon.proc.pid, signal.SIGHUP)
         self.assertTrue(daemon.isAlive())
 
-    @test_base.flaky
+    #@test_base.flaky
```
Comment (theopolis) [inline tools/tests/test_osqueryd.py]: Same.

**Comment by osqueryer** [comment]
Build origin/pr/1949/merge (Job: 2505) is good! :+1:

**Comment by osqueryer** [comment]
Build origin/pr/1949/merge (Job: 2506) is good! :+1:

File: tools/tests/test_osqueryi.py [hunk 7]
```diff
@@ -50,6 +51,7 @@ def test_config_check_success(self):
         self.assertEqual(proc.stdout, "")
         print(proc.stdout)
         print(proc.stderr)
+        time.sleep(1)
         self.assertEqual(proc.proc.poll(), 0)
```
Comment (theopolis) [inline tools/tests/test_osqueryi.py]: Remove this

File: osquery/core/init.cpp [hunk 8]
```diff
@@ -107,41 +104,51 @@ volatile std::sig_atomic_t kHandledSignal{0};
...
-      // Re-raise the handled signal, which has since been restored to default.
-      raise(signal);
+
+    // Handle signals based on a tri-state (worker, watcher, neither).
+    if (num == SIGHUP) {
+      if (!isWatcher() || hasWorkerVariable()) {
+        // Reload configuration.
+      }
+    } else if (num == SIGTERM || num == SIGINT || num == SIGABRT ||
+               num == SIGUSR1) {
```
Comment (theopolis) [inline osquery/core/init.cpp]: Is this signal handler installed?

**Comment by osqueryer** [comment]
Build origin/pr/1949/merge (Job: 2507) is good! :+1:

**Pull request merged**